cpu/lm4f120/periph/timer.c

```diff
@@ -27,132 +29,384 @@
+{
+    const unsigned long long scaledv = (unsigned long long) uncorrected * diviser;
+    return scaledv;
+}
+
+static inline unsigned int _llvalue_to_scaled_value(unsigned long long corrected, unsigned int diviser)
+{
+    const unsigned long long scaledv = corrected / diviser;
+    return scaledv;
+}
+
 int timer_init(tim_t dev, unsigned int us_per_tick, void (*callback)(int))
 {
-    if (dev == TIMER_0) {
-        config[dev].cb = callback;                          /* User Function */
-        ROM_SysCtlPeripheralEnable(SYSCTL_PERIPH_WTIMER0);  /* Activate Timer0 */
-        WTIMER0_CTL_R &= ~0x00000001;                       /* Disable timer0A during setup */
-        WTIMER0_CFG_R  = TIMER_CFG_16_BIT;
-        WTIMER0_TAMR_R = TIMER_TAMR_TAMR_PERIOD;            /* Configure for periodic mode */
-        WTIMER0_TAPR_R = TIMER_0_PRESCALER;                 /* 1us timer0A */
-        WTIMER0_ICR_R  = 0x00000001;                        /* clear timer0A timeout flag */
-        WTIMER0_IMR_R |= 0x00000001;                        /* arm timeout interrupt */
-        ROM_IntPrioritySet(INT_WTIMER0A, 32);
-        timer_irq_enable(dev);
-        timer_start(dev);
-        DEBUG("startTimeout Value=0x%lx\n", ROM_TimerValueGet(WTIMER0_BASE, TIMER_A));
-        return 1;
-    }
-    return -1;
+    if (dev >= TIMER_NUMOF){
+        return -1;
+    }
+
+    config[dev].cb = callback;                          /* User Function */
+    config[dev].diviser = us_per_tick * ROM_SysCtlClockGet()/1000000;
+
+    unsigned int sysctl_timer;
+    unsigned int timer_base;
+    unsigned int timer_side = TIMER_A;
+    unsigned int timer_cfg = TIMER_CFG_SPLIT_PAIR | TIMER_CFG_A_PERIODIC_UP;
+    unsigned int timer_max_val;
+    unsigned int timer_intbit = TIMER_TIMA_TIMEOUT | TIMER_TIMA_MATCH;
+
+    switch(dev){
+#if TIMER_0_EN
+    case TIMER_0:
+        sysctl_timer = SYSCTL_PERIPH_WTIMER0;
+        timer_base = WTIMER0_BASE;
+        timer_max_val = TIMER_0_MAX_VALUE;
+        break;
+#endif
+#if TIMER_1_EN
+    case TIMER_1:
+        sysctl_timer = SYSCTL_PERIPH_WTIMER1;
+        timer_base = WTIMER1_BASE;
+        timer_max_val = TIMER_1_MAX_VALUE;
+        break;
+#endif
+    default:
+        return -1; /* unreachable */
+    }
+
+    if (timer_side == TIMER_A) {
+        timer_cfg |= TIMER_TAMR_TAMIE;
+    }
+    else if (timer_side == TIMER_B) {
+        timer_cfg |= TIMER_TBMR_TBMIE;
+    }
+
+    ROM_SysCtlPeripheralEnable(sysctl_timer);
+
+    ROM_TimerDisable(timer_base, timer_side);
+    ROM_TimerConfigure(timer_base, timer_cfg);
+
+    unsigned long long lltimer_val_max = _scaled_to_ll_value(timer_max_val, config[dev].diviser);
+
+    ROM_TimerPrescaleSet(timer_base, timer_side, lltimer_val_max >> 32);
+    ROM_TimerLoadSet(timer_base, timer_side, lltimer_val_max & 0xFFFFFFFF);
+    ROM_TimerIntClear(timer_base, timer_intbit);
+
+    ROM_TimerIntEnable(timer_base, timer_intbit);
+
+    timer_irq_enable(dev);
+    timer_start(dev);
+
+    return 1;
```

Comment:
Yes, but this is `timer_init` right? [API Doc Link](http://riot-os.org/api/group__drivers__periph__timer.html#ga270861eb8467e0eafe1d9cf56483362c)